### PR TITLE
Add new APIs for safer mount specifications

### PR DESCRIPTION
**Pull request by cpuguy83**
This is for docker/docker#22373

### CODE REVIEW

**Comment by cpuguy83** [comment]
ping @thaJeztah @justincormack

**Comment by cpuguy83** [comment]
omg `go lint`....

**Comment by MHBauer** [comment]
Remove the volume structs from the base `types` package as well, or leave it there for compatibility reasons? Clients will have to change to the new package.

**Comment by cpuguy83** [comment]
IIRC I moved these around due to circular imports.

**Comment by cpuguy83** [comment]
Yes, so issue is `types/container` pkg needs volume create struct, which caused an import to `types` pkg, and `types` is already importing `types/container`... so the volume `CreateRequest` needs to be in a separate package.
Automatically I'm going to bring over `ListResponse` to that new package b/c these should not be split up, which means I also need to bring over the `Volume` struct or else have another circular import... `types/container` -> `types/volume` -> `types`

**Comment by cpuguy83** [comment]
Lint issues and all are fixed, CI is passing now.

File: types/container/host_config.go [hunk 1]
```diff
@@ -317,4 +334,7 @@ type HostConfig struct {
 
 	// Contains container's resources (cgroups, ulimits)
 	Resources
+
+	// Mounts sepcs used by the container
```
Comment (thaJeztah) [inline types/container/host_config.go]: s/sepcs/specs/

**Comment by thaJeztah** [comment]
LGTM

**Comment by calavera** [comment]
@cpuguy83 this is modifying the existing endpoints, does that mean that the docker cli is going to use them too?

**Comment by cpuguy83** [comment]
@calavera Not sure I follow.

**Comment by calavera** [comment]
oh, nevermind, I should have read more closely.

**Comment by cpuguy83** [comment]
This is now updated and uses (or rather moves) the swarm mount types.

**Comment by thaJeztah** [comment]
re-LGTM

File: types/types.go [hunk 2]
```diff
@@ -410,13 +411,14 @@ type DefaultNetworkSettings struct {
 
 // MountPoint represents a mount point configuration inside the container.
 type MountPoint struct {
-	Name        string `json:",omitempty"`
+	Type        mount.Type `json:",omitempty"`
+	Name        string     `json:",omitempty"`
 	Source      string
 	Destination string
```
Comment (stevvooe) [inline types/types.go]: Wish we could alias these fields. I'm guessing this is exported via docker info or docker inspect?

**Comment by stevvooe** [comment]
LGTM

I like that we've combined these.

File: types/types.go (hunk 2 above)
Comment (cpuguy83) [inline types/types.go]: Yes.

**Comment by cpuguy83** [comment]
With 2 LGTMs I'm merging.
Thanks!

**Pull request merged**